### PR TITLE
feat(docs-sync): learn from maintainer corrections

### PR DESCRIPTION
**Pull request by iscekic**
## Summary

Add learning from maintainer corrections to the rolling docs pull request.

The workflow reads trusted corrective commits and review comments. It extracts validated rules, injects tagged rules into triage and edit, and commits `LEARNINGS.md` with the normal rolling pull request update.

## Design and safety

The implementation keeps the twelve closed design decisions and G5.

- G1 reads standalone review comments.
- G2 trusts only OWNER, MEMBER, and COLLABORATOR authors.
- G3 allowlists `LEARNINGS.md` in the non-content draft gate.
- G4 accepts the barren-run lag.
- G5 PATCHes `learned-through` for empty deltas and uses the upsert route for non-empty deltas.

The extractor rejects rules that name a pull request, page, person, or URL. It preserves a marker when extraction omits its output. It writes no marker under `dry_run` or `LEARNINGS_NO_PATCH`.

## Upsert changes

- `nonContentFiles` excludes the root learnings file from the draft gate.
- `renderBody` renders `learned-through` only when a value exists.
- `resolveLearnedThrough` preserves an existing marker when extraction omits an output.
- A newly created rolling pull request is assigned to @emilieschario and asks that user for a review. A failure there warns and does not fail the run.

## Local proof

The extraction call deliberately omits `--auto`. The local proof used `LEARNINGS_NO_PATCH=1`, the real `9dd2c07` correction, and `kilo/moonshotai/kimi-k3`.

It produced: `Do not document features that are behind experimental or unreleased flags.`

- Scope: `both`
- Source: `commit:9dd2c07`

`kilo-auto/efficient` was unavailable locally. The workflow model completed the proof with no live pull request write.

## Review result

The planning reviews raised 31 findings. The plan accepted 29 and rejected two.

The rejected items were a prose-arithmetic budget assertion and a persistent learning tombstone. Both add fragile or unsupported state without preventing a real defect.

## Accepted limits

The barren-run lag remains accepted. The conflict path can reset the marker and self-heals on the next run.

## Visual Changes

N/A

## Remaining: one human-approved dispatch

- What remains: run `workflow_dispatch` for `docs-sync` on this branch.
- It can update live pull request #12716 on `docs/auto-sync`.
- It can add a `LEARNINGS.md` commit and advance `learned-through`.
- Run first with `dry_run: true`.
- The dry run proves extraction without a file or marker write.
- Then check the marker, one rule per line, and rule text without pull requests, pages, people, or URLs.

E2E: skipped — this change is a GitHub Action and has no local runtime surface.

### CODE REVIEW

File: .github/docs-sync/learn.mjs [hunk 1]
```diff
@@ -0,0 +1,885 @@
...
+
+// --- helpers for main ---
+
+function git(args) {
+  return execFileSync("git", args, { stdio: ["ignore", "pipe", "inherit"] })
+    .toString()
+    .trim()
+}
+
+function sorted(list) {
```
[the remark this answers is not in the document]
Comment (iscekic) [reply .github/docs-sync/learn.mjs]: (bot) Removed the unused sorted helper.

File: .github/docs-sync/learn.mjs [hunk 2]
```diff
@@ -0,0 +1,885 @@
...
+        author_association: c.author_association,
+        path: c.path,
+        body: c.body ?? "",
+      }
+    } else {
+      candidates.push({
+        source: `comment:${c.id}`,
+        date: (c.created_at ?? "").slice(0, 10),
+        path: c.path,
+        body: c.body ?? "",
```
[the remark this answers is not in the document]
Comment (iscekic) [reply .github/docs-sync/learn.mjs]: (bot) Capped comment bodies at 5,000 characters before extraction input.

File: .github/docs-sync/learn.mjs [hunk 3]
```diff
@@ -0,0 +1,885 @@
...
+
+  // Step 12: route by outcome (G5 table, exact).
+  if (nonEmpty) {
+    // Non-empty validated delta.
+    const newEntries = applyDelta(existing, { add: validated.add, remove: validated.remove })
+    fs.writeFileSync(`${OUT_DIR}/learnings.json`, JSON.stringify(newEntries, null, 2))
+    const marker = renderLearnedThrough({ commit: tipSha, comment: maxCommentAt })
+    const suppressed = process.env.DRY_RUN === "true" || process.env.LEARNINGS_NO_PATCH === "1"
+    if (!suppressed) appendOutput("learned_through", marker)
+    if (suppressed) log(`learned-through output suppressed: ${marker}`)
```
[the remark this answers is not in the document]
Comment (iscekic) [reply .github/docs-sync/learn.mjs:750]: (bot) Added non-empty suppression coverage for both safety switches.

File: .github/docs-sync/learn.mjs [hunk 4]
```diff
@@ -0,0 +1,889 @@
...
+  })
+
+  // Step 7: correlate comments to commits.
+  // A comment is a commit's trigger when c.path is in that commit's full file list
+  // and c.created_at < commit date. The earliest such commit claims it.
+  for (const c of trustedComments) {
+    let best = null
+    for (const cc of candidates) {
+      if (!Array.isArray(cc.files) || !cc.files.includes(c.path)) continue
+      if (c.created_at < cc.iso) {
```
[the remark this answers is not in the document]
Comment (iscekic) [reply .github/docs-sync/learn.mjs]: (bot) Parsed timestamps before correlation and added offset coverage.

File: .github/docs-sync/learn.mjs [hunk 5]
```diff
@@ -0,0 +1,892 @@
...
+      // Required for the first live run: the rolling branch predates the seeded file.
+      try {
+        existingText = git(["show", `origin/main:${LEARNINGS_FILE}`])
+      } catch {
+        existingText = ""
+      }
+    }
+    existing = parseLearnings(existingText)
+  }
+  log(`existing entries: ${existing.length}`)
```
Comment (pandemicsyn) [inline .github/docs-sync/learn.mjs:451]: Extraction errors can disable all existing learnings for the run

Suggested fix: Emit prompt artifacts from the best available existing `LEARNINGS.md` before any fallible API/model work, and retain them on every error path. Ideally pre-seed from the checked-out file at startup, then replace them after loading the rolling branch and after a validated delta. Add a test where PR/comment fetching throws and assert triage/edit still receive existing rules.

File: .github/docs-sync/learn.mjs [hunk 6]
```diff
@@ -0,0 +1,892 @@
...
+  if (fixture) {
+    // Write to the fixture patch file instead of the network.
+    fs.writeFileSync(patchFile, marker)
+    log(`wrote marker to ${patchFile}`)
+    return
+  }
+
+  // Live PATCH: body-only, one line changed. The job already holds pull-requests: write.
+  const { api, repo } = await import("./lib.mjs")
+  const newBody = patchMarkerIntoBody(prBody, marker)
```
Comment (pandemicsyn) [inline .github/docs-sync/learn.mjs:815]: The direct marker PATCH can overwrite concurrent PR-body edits

Suggested fix: Fetch the PR again immediately before patching, apply `patchMarkerIntoBody` to that latest body, and PATCH the result. Add a test in which the body returned at patch time differs from the initially fetched body and verify the intervening content is preserved. If GitHub exposes a usable conditional-update mechanism, use it to close the remaining fetch/PATCH race.

File: .github/docs-sync/learn.mjs [hunk 7]
```diff
@@ -0,0 +1,892 @@
...
+      reason = "rule text absent, shorter than 10 characters, or longer than 300"
+    } else if (!["triage", "edit", "both"].includes(a.scope)) {
+      reason = `invalid scope: ${a.scope}`
+    } else if (!/^commit:[0-9a-f]{7,40}$/.test(a.source) && !/^comment:\d+$/.test(a.source)) {
+      reason = `invalid source format: ${a.source}`
+    } else if (!candidates.includes(a.source)) {
+      reason = `source ${a.source} not in candidate sources`
+    } else if (!/^[a-z0-9][a-z0-9-]{2,48}$/.test(a.id)) {
+      reason = `invalid id format: ${a.id}`
+    } else if (existingIds.has(a.id) && !toRemove.includes(a.id)) {
```
Comment (pandemicsyn) [inline .github/docs-sync/learn.mjs:238]: Duplicate IDs within one model response pass validation

Suggested fix: Track IDs accepted from the current `add` array as well as existing IDs, rejecting a second occurrence. Also compare normalized text against already accepted additions so one response cannot render duplicate rules. Add validation tests for duplicate IDs and duplicate rule text in the same delta.

**Review by pandemicsyn — APPROVED** [review]
couple small things inline from the robot
